man/man1/lineid.1
.TH LINEID 1NEMO "19 January 2026"

.SH "NAME"
lineid \- spectral line ID tool

.SH "SYNOPSIS"
\fBlineid\fP in=file [parameter=value]

.SH "DESCRIPTION"
\fBlineid\fP is a tool that helps with converting between frequencies and doppler velocities, as well
as identifying lines when the rest frequency can be derived.

.PP
\fBlineid\fP currently operates in three modes:

.br
1. For a given doppler velocity (\fBvel=\fP) and an input list of sky
frequencies, the rest frequencies are calculated, and if a linelist
is given, a lineid is attempted. This will be useful for multiple
spectral lines in a given object.

.br
2. For a given rest frequency (\fBrestfreq=\fP) and an input list of
sky frequencies, their doppler shifts are calculated in a variety of
conventions (radio, relativistic, optical, and z). This will be useful
for a large field of view where always the same line (e.g. HI) is
detected in many sources at different cosmological distances.

.br
3. For a given \fBrestfreq=\fP and an input list of velocities (now
with \fBxunit=km/s\fP), the rest frequencies of the other lines are
computed, and a lineid can be attempted. The first velocity in the
list needs to match the given rest frequency This is probably a rare
case, as spectra typically originate in frequency, not doppler
velocity.

.br
4. For given \fBvel=\fP and list of velocities there is no
practical application? It can probably be run as mode 3.

.PP
In cases 1. and 3. a lineid can be attempted if a line list is supplied using \fBlinelist=\fP.
See LINELIST below for examples of line lists.

.PP
\fBlineid\fP can also attempt to find the strongest peak
in a spectrum if both the X and Y columns are given, but
\fItabpeak(1NEMO)\fP is probably a better feed into \fBlineid\fP.

.PP
More sophisticated algorithms for lineid can be found in ADMIT's LineID task, in particular
dealing with multiple "spectra" derived from a fits cube.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP\fIfile\fP
Input file name, in ASCII \fItable(5NEMO)\fP format.  The number of columns will depend
on which of the xcol,ycol,dycol is used.
.br
No default.
.TP
\fBxcol=\fP\fIx-column\fP
X coordinate column, representing the frequency or wavelength axis.
Units are GHz, unless specified with the \fBxunit\fP keyword. The X column
is always required.
.br
[1]
.TP
\fBycol=\fP\fIy-column\fP
Y coordinate column, reprepenting an intensity. If not used, the xcol is assumed
to have the already determined values of the lines to be identified or
converted to doppler velocities.
.br
By default not used.
.TP
\fBdxcol=\fP\fIx-err-column\fP
Errors in the X coordnates. Only makes sense if no ycol is given.  Not implemented yet.
.br
By default not used.
.TP
\fBxunit=\fP
X axis unit. Currently only GHz and km/s are supported.
.br
[GHz]
.TP
\fBvel=[#[,unit]]\fP
Redshift of object, in km/s. Optional unit added, e.g. \fIvel=1.2,z\fP.
As for unit, only \fIkm/s\fP and the dimensionless \fIz\fP are allowed.
.br
Not set by default. If a value is given, the unit defaults to \fIkm/s\fP.
.TP
\fBrestfreq=\fP
If line is known, give rest frequency in \fBxunit\fP (only GHz supported now).
In a future version \fIrestwave=\fP might be supported.
.br
Not set by default.
.TP
\fBlinelist=\fP
Table with frequencies and line names. There is a simple list in $NEMODAT/z_lines.list
with some common high-Z lines.
.br
Not set by default.
.TP
\fBdv=\fP
Report all lines within \fBdv\fP km/s.
Negative values will use nearest neighbor and report the dv, and report but comment
out the line if the deviation is larger than -dv.
.br
[-10]
.TP
\fBclip=\fP
Do not fit peaks in the spectrum below this clip level.
.br
Default: not used
.TP
\fBveldef=RAD|OPT|REL\fP
Mode of velocity axis. Radio, Relativistic, or Optical are supported.
.br
[OPT]

.SH "CAVEATS"

The code was developed with frequency in mind, only GHz and km/s are supported. Wavelength
is planned.

Looks for a single emission line in the strongest peak.. May not work for absorption lines yet, or
use \fItabpeak(1NEMO)\fP if multiple lines are needed. See EXAMPLES below.

Although different doppler conventions are supported, coordinate frames are ignored.

.SH "EXAMPLES"
The following three example generates a spectrum from a random table of 115 values,
smoothes by 10 channels to create a handful
of peaks and investigates the strongest peak, which happens to be near 40 "GHz".  The frequencies
are labeled 0 to 114 in steps of 1 GHz.
.EX

  #> mode 1.
  tabgen - 115 1 2 123 | tabsmooth - 1 12 10,5 |\\
        lineid - 0 1 vel=1.872,z dv=200 linelist=$NEMODAT/z_lines.list
  ### nemo Debug[0]: Vrad=195408 Vrel=234961 Vopt=561211 z=1.872
  ### nemo Debug[0]: Found 23 lines in the linelist=/home/teuben/nemo/data/z_lines.list
  # There are 1 restfreq's to identify:
  # Commented lines do not fall within dv=200 km/s
  #    guess    ID    LINE     dFreq    deltaV  NAME
  #    GHz            Ghz      GHz      km/s
    115.230827   2 115.271202 0.040375 105.006 CO(1-0)


  #> mode 2.
  tabgen - 115 1 2 123 | tabsmooth - 1 12 10,5 | tabpeak -  0 1 |\\
        lineid - restfreq=115
  # restfreq=115
  # SkyFreq Vrad  Vrel   Vopt    z
  # GHz     km/s  km/s  km/s    
  29.914   221810 261794 852717  2.84436
  40.1222  195198 234729 559487  1.86625
  72.1352  111744 130493 178145  0.594229


  #> mode 3.
  tabgen - 115 1 2 123 | tabsmooth - 1 12 10,5 | tabpeak -  0 1 |\\
        lineid - restfreq=115 xunit=km/s
  ### nemo Debug[0]: Reading 2 column(s)
  # There are 3 restfreq's to identify: (use linelist= to find their names)
  115.000000
  114.996085
  114.983808

.EE

and here some examples with no input spectrum. Only \fBxcol=\fP is needed:

.EX

  #> mode 1.
  echo 1.397105 | lineid - vel=5000 
  ### nemo Debug[0]: Vrad=4917.98 Vrel=4958.31 Vopt=5000 z=0.0166782
  # There are 1 restfreq's to identify: (use linelist= to find their names)
  1.420406
  

  #> mode 2.
  nemoinp 1.397105,1.1 | lineid - restfreq=1.420405751786 
  # restfreq=1.42041
  # SkyFreq Vrad    Vrel     Vopt     z
  # GHz     km/s    km/s     km/s
  1.39711  4916.83  4957.14  4998.81  0.0166742
  1.1     67625.2  75009.8  87322.9   0.291278


  #> mode 3.
  echo 5000 | lineid - restfreq=1.420405751786 xunit=km/s
  ### nemo Debug[1]: 0a: Line at 5000.000000 km/s has skyfreq 1.397105  restfreq 1.420406
  # There are 1 restfreq's to identify: (use linelist= to find their names)
  1.420406

.EE

An contrived example of the CO(1-0) and CO(2-1) lines from the CO
ladder, deliberately doppler shifted a bit by 300 km/s (give or take):

.EX

  #> mode 1.
  nemoinp 110.092,115.155,220.179,230.308  |\\
      lineid - linelist=$NEMODAT/z_lines.list vel=300 dv=2.5
  ### nemo Debug[0]: Vrad=299.7 Vrel=299.85 Vopt=300 z=0.00100069
  ### nemo Debug[0]: Found 23 lines in the listlist /home/teuben/NEMO/nemo/data/z_lines.list
  # There are 4 restfreq's to identify:
  #    guess    ID    LINE     dFreq    deltaV   NAME
  #    GHz            Ghz      GHz      km/s
    110.202168   1 110.201354 0.000814 -2.21418  13CO(1-0)
  # 115.270235   2 115.271202 0.000967  2.51514  CO(1-0)
    220.399331   3 220.398680 0.000651 -0.886089 13CO(2-1)
    230.538467   4 230.538000 0.000467 -0.607859 CO(2-1)
    
.EE

Here an example of detected lines in velocity space and deriving their
restfreq's.  The first line in the list will be the one belonging to
the known restfreq of CO(1-0), and was detected at -20 km/s. But there
was a weak line at -312 km/s with unknown identification. To find what
line this is, We need to find the rest frequency and then a lineid. In
the example we only grab the restfreq's.

.EX

  #> mode 3.
  nemoinp -20,-310 | lineid - restfreq=115.2712018 xunit=km/s
  # There are 2 restfreq's to identify: (use linelist= to find names)
  115.271202
  115.382823
  
.EE

There is indeed a known SiC2,v=0 line at 115.382367 GHz.

.PP

Compare different doppler conventions (veldef)
.EX

  #> mode 1.
  echo 0 | lineid - vel=5000 veldef=rel
  ### nemo Debug[0]: rad=4958.99 rel=5000 opt=5042.4 z=0.0168196

.EE

.SH "LINELIST"
The input linelist is a very simple ascii table of (frequency,name).
There are many sources for a linelist, we mention a few:
.nf

1. $NEMO/data/z_lines.list : common high-Z lines, from HI at 1.4GHz upwards

2. ADMIT'S has a few in $ADMIT/etc/*list

3. Splatalogue at https://www.splatalogue.online/

4. Astroquery:   https://astroquery.readthedocs.io/en/latest/nist/nist.html and
   https://astroquery.readthedocs.io/en/latest/atomic/atomic.html


.fi

.SH "SEE ALSO"
tablsqfit(1NEMO), tabpeak(1NEMO), tabpp(1NEMO), ccdsky(1NEMO), table(5NEMO)
.PP
admit::LineID_AT()

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.25i +4.5i
13-dec-2024	0.1 Draft created	PJT
3-jul-2025	0.3 add mode=	PJT
15-jan-2026	0.4 added simple nearest neighbor line_id	PJT
16-jan-2026	0.5 implemented veldef=		PJT
17-jan-2026	0.6 mode= deprecated	PJT
18-jan-2026	0.7 removed never used keywords		PJT
.fi


src/kernel/tab/lineid.c
/*
 *  LINEID:   spectral line ID tool
 *
 *    13-dec-2024   0.2 simple brightest peak finder.
 *     3-jul-2025   0.3 add mode=
 *    15-jan-206    0.4 all modes now work, added simple nearest neighbor line_id
 *
 *
 *  Here's a contrived example of taking some lines, rounding them to 2 digits, and fitting them to the same linelist
 
 tabcols $NEMODAT/z_lines.list 1 | tabmath - - '%1/(1+0.001)' all format=%.2f | lineid - mode=0 linelist=$NEMODAT/z_lines.list vel=300 dv=2.5

 */

/**************** INCLUDE FILES ********************************/ 
 
#include <stdinc.h>
#include <getparam.h>
#include <extstring.h>
#include <axis.h>
#include <table.h>
#include <mdarray.h>
#include <moment.h>
#include <mks.h>
#include <lsq.h>


/**************** COMMAND LINE PARAMETERS **********************/

string defv[] = {                /* DEFAULT INPUT PARAMETERS */
  "in=???\n              Input (table) file name",
  "xcol=1\n		 x coordinate column of spectrum or linesq", 
  "ycol=\n		 y coordinate column of spectrum",
  "dxcol=\n              If used, columns with errors in the x coordinate [not implemented]",
  "xunit=GHz\n           x-axis unit (GHz, km/s)",
  "vel=\n                VEL of object, if known (km/s)",
  "restfreq=\n           RESTFREQ (or RESTWAVE?) in xunits, if known",
  "linelist=\n           ASCII linelist (freq,label), e.g. $NEMODAT/z_lines.list",
  "dv=-1\n               Slop in velocity to allow in line_id (km/s)  -1 means nearest match",
  "clip=\n               Do not fit peaks below this clip level",
  "veldef=OPT\n          doppler_convention: OPT, RAD or REL",
  "VERSION=0.8\n	 19-jan-2026 PJT",
  NULL
};

string usage = "spectral line ID tool";


/**************** GLOBAL VARIABLES ************************/

local string input;				/* filename */
local stream instr;				/* input file */
local table *tptr;                              /* table */
local mdarray2 d2;                              /* data[col][row] */

local int xcol, ycol, dxcol;
local int mode = -1;                            /* 0: lines known, just need to ID   1: spectrum given */

local real *rfreq;                              /* restfreq estimated go here */
local int nrfreq;                               /* active length of this array */
local real *dvel ;                              /* doppler velocity estimates */
local int ndvel;                                /* active length of this array */

local real  *x, *y;    			        /* data from file */
local int    npt;				/* actual number of data points */

local bool Qvlsr, Qrest;
local real vlsr;
local real restfreq;
local real dv;

local string linelist = NULL;                   /* linelist (must be file, cannot be pipe) */
local real   *lfreq;                            /* linelist freqs */
local string *lname;                            /* linelist names */
local int    lnum = 0;                          /* #lines in linelist */

#define MAXU 64
local string xunit;
local int xunit_mode = 0;
local char vlsr_unit[MAXU];
local string veldef_s;
local int veldef;

#define MAXCOL    2
#define ckms 299792.458


#define UNIT_MHZ 1
#define UNIT_GHZ 2
#define UNIT_A   3
#define UNIT_NM  4
#define UNIT_KMS 6
#define UNIT_Z   7

#define VEL_RAD  1
#define VEL_REL  2
#define VEL_OPT  3


void set_params(void);
void read_data(void); 
void peak_data(void); 
void line_id(void);
void do_peak(real *xpeak, real *xerr, real *ypeak);
void get_nu(string kv, real *value, string unit, string defunit);
  
void nemo_main()
{
  set_params();
  read_data();
  peak_data();
  line_id();
}

/*
 *     get_nu :   parse a   NUMBER,UNIT    string
 *     input:   kv         e.g.   "1,pc"
 *              defunit    e.g.   "pc"
 *     output:  value      
 *              unit
 *
 */

void get_nu(string kv, real *value, string unit, string defunit)
{
  string *sp = burststring(kv,",");
  int nsp = xstrlen(sp,sizeof(string)) - 1;
  int nv;

  *value = 1.0;
  strcpy(unit,defunit);
  if (nsp == 0) return;

  nv = nemoinpr(sp[0],value,1);
  if (nv != 1) error("error parsing %s",sp[0]);
  if (nsp == 1) return;

  strcpy(unit,sp[1]);
}

// radio to optical
inline real vrad2vopt(real vrad) {    return vrad/(1-vrad/ckms); }
// optical to radio
inline real vopt2vrad(real vopt) {    return vopt/(1+vopt/ckms); }
// optical to z
inline real vopt2z(real vopt) {       return vopt/ckms; }
// z to optical
inline real z2vopt(real z) {          return ckms*z; }
// z to radio
inline real z2vrad(real z) {          return vopt2vrad( z2vopt(z) ); }
// radio to z
inline real vrad2z(real vrad) {       return vopt2z( vrad2vopt(vrad) ); }
// relativistic to z
inline real vrel2z(real vrel) {       real b = vrel/ckms;          return sqrt((1+b)/(1-b)) - 1.0; }
// z to beta
inline real z2beta(real z) {          real z1=1.0+z;   z1=z1*z1;   return (z1-1)/(z1+1); }
// z to relativistic
inline real z2vrel(real z) {          return z2beta(z) * ckms; }



void set_params()
{
  input = getparam("in");             /* input table file */
  instr = stropen (input,"r");
  tptr = table_open(instr,0);

  xcol = getiparam("xcol");
  if (hasvalue("ycol")) {
    ycol = getiparam("ycol");
    mode = 1;
  } else {
    mode = 0;
  }

  if (hasvalue("dxcol"))
    warning("dxcol= not supported yet");

  dv = getdparam("dv");

  if (hasvalue("linelist"))
    linelist = getparam("linelist");

  xunit = getparam("xunit");
  if (streq(xunit,"GHz"))  xunit_mode = UNIT_GHZ;
  if (streq(xunit,"km/s")) xunit_mode = UNIT_KMS;
  if (xunit_mode == 0) error("xunit %s not supported yet",xunit);

  veldef_s = getparam("veldef");
  veldef = 0;
  if (strncmp(veldef_s, "opt", 1) == 0) veldef = VEL_OPT;
  if (strncmp(veldef_s, "rad", 2) == 0) veldef = VEL_RAD;
  if (strncmp(veldef_s, "rel", 2) == 0) veldef = VEL_REL;
  if (strncmp(veldef_s, "OPT", 1) == 0) veldef = VEL_OPT;
  if (strncmp(veldef_s, "RAD", 2) == 0) veldef = VEL_RAD;
  if (strncmp(veldef_s, "REL", 2) == 0) veldef = VEL_REL;
  if (veldef==0) error("veldef=%s not a valid option",veldef_s);
  dprintf(1,"veldef=%d\n",veldef);

  Qvlsr = hasvalue("vel");
  Qrest = hasvalue("restfreq");
  if (Qvlsr) {
    get_nu(getparam("vel"), &vlsr, vlsr_unit, "km/s");
    if (streq(vlsr_unit,"z")) {
      if (veldef != VEL_OPT) warning("You used z=%g, but did not specify veldef=optical convention",vlsr);
      vlsr *= ckms;
    }
    if (veldef == VEL_REL) {
      if (vlsr > ckms) error("Bad vlsr for veldef=RELATIVISTIC");
      vlsr = z2vopt(vrel2z(vlsr));
    }
    if (veldef == VEL_RAD) {
      if (vlsr > ckms) error("Bad vlsr for veldef=RADIO");      
      vlsr = vrad2vopt(vlsr);
    }
    dprintf(0,"Vrad=%g Vrel=%g Vopt=%g z=%g\n",vopt2vrad(vlsr),z2vrel(vopt2z(vlsr)),vlsr,vopt2z(vlsr));
  }
  if (Qrest) restfreq = getdparam("restfreq");
  if (Qvlsr && Qrest) {
    if (xunit_mode == UNIT_KMS) {
      warning("Mode not implemented yet");
    } else
      error("Cannot give both vel= and restfreq= yet");
  }

}

void read_data()
{
  int colnr[1+MAXCOL];
		
  dprintf(2,"Reading datafile, xcol,ycol=%d..,%d,...\n",xcol,ycol);
    
  colnr[0]  = xcol;
  colnr[1]  = ycol;
  d2 = table_md2cr(tptr, 2, colnr,0,0);
  npt = table_nrows(tptr);
  dprintf(1,"Found %d rows\n", npt);
  
  x = &d2[0][0];
  y = &d2[1][0];

  // get minmax in X and Y

  // code will populate either of these two arrays
  rfreq  = (real *) allocate(npt * sizeof(real));
  dvel   = (real *) allocate(npt * sizeof(real));
  nrfreq = 0;
  ndvel  = 0;

  // manually parse the linelist, only look at first and second column
  // (freq,name)

  if (linelist != NULL) {
    char line[MAX_LINELEN];
    real fmin, fmax;
    string *words;
    int nw;
    int nmax = nemo_file_lines(linelist, 0);
    dprintf(1,"nemo_file_lines: %d\n",nmax);
    lfreq = (real *) allocate(nmax*sizeof(real));
    lname = (string *) allocate(nmax*sizeof(string));

    instr = stropen(linelist,"r");
    while (fgets(line, MAX_LINELEN, instr) != NULL) {
      if (line[0] == '#') continue;
      words = burststring(line," \n");
      nw = xstrlen(words,sizeof(string))-1;
      dprintf(1,"%d  %f %s %d\n", lnum, atof(words[0]), words[1], nw);
      lfreq[lnum] = atof(words[0]);
      lname[lnum] = strdup(words[1]);
      if (lnum==0) {
	fmin = fmax = lfreq[0];
      } else {
	fmin = MIN(fmin, lfreq[lnum]);
	fmax = MAX(fmax, lfreq[lnum]);
      }
      lnum++;
    }
    strclose(instr);
    dprintf(0,"Found %d lines in the linelist=%s; range %f to %f\n",lnum,linelist,fmin,fmax);
  }
}

void peak_data()
{
  real xpeak, xerr, ypeak;
  real v, z, z1, z2, rf2, sf2;
  int i;
  
  dprintf(1,"xunit=%s\n",xunit);

  if (mode == 1 ) {
    //warning("Lines from peak fits in table:");    
    // for now: simple brightest peak finder
    do_peak(&xpeak, &xerr, &ypeak);
  } else if (mode == 0) {
    //warning("Direct line from table:");
    xpeak = x[0];
    ypeak = y[0];
  } else
    error("mode=%d not implemented", mode);

  if (mode == 1) {  // peak fit to (xcol,ycol)
    if (Qrest) {
      z = restfreq/xpeak - 1;
      v = z*ckms;
      dprintf(1,"1a: Line at %f %s has z=%g or vel=%g km/s (cz)\n",xpeak,xunit,z,v);
      // add vel
      dvel[ndvel++] = v;
      // add skyfreq in the restfreq slot
      rfreq[nrfreq++] = xpeak;
    } else if (Qvlsr) {
      z = vlsr/ckms;
      restfreq = xpeak * (1+z);
      dprintf(1,"1b: Line at %f, Look near freq = %f %s for a lineid; vel=%g\n", xpeak, restfreq, xunit,vlsr);
      // add freq
      rfreq[nrfreq++] = restfreq;
    } else {
      dprintf(0,"Peak: %g %g\n", xpeak, ypeak);
      //rfreq[nrfreq++] = xpeak;
    }
  } else { // mode=0:    (xcol has freq/vel; ycol not used)
    if (xunit_mode == UNIT_KMS) {
      if (Qrest) {
	for (i=0; i<npt; i++) {
	  if (veldef == VEL_OPT)
	    z2 = x[i]/ckms;
	  else if (veldef == VEL_RAD)
	    z2 =  vrad2z(x[i]);
	  else if (veldef == VEL_REL)
	    z2 =  vrel2z(x[i]);
	  sf2 = restfreq/(1+z2);
	  if (i==0) // first one is reference 
	    z1 = z2;
	  rf2 = sf2 * (1+z1);
	  dprintf(1,"0a: Line at %f %s has skyfreq %f  restfreq %f\n",x[i],xunit,sf2,rf2);
	  // add freq
	  rfreq[nrfreq++] = rf2;
	}
      } else if (Qvlsr) {
	//warning("not implementable");
	return;
      }
    } else if (xunit_mode == UNIT_GHZ) {
      if (Qrest) {
	for (i=0; i<npt; i++) {
	  z2 = restfreq/x[i] - 1.0;
	  v = z2*ckms;
	  dprintf(1,"0b: Line at %f %s has z=%f or vel=%f km/s (cz)\n",x[i],xunit,z2,v);
	  // add vel
	  dvel[ndvel++] = v;
	}
      } else if (Qvlsr) {
	for (i=0; i<npt; i++) {
	  // convert to opt if not opt
	  if (veldef == VEL_OPT)
	    z2 = vlsr/ckms;
	  else if (veldef == VEL_RAD)
	    z2 = vrad2z(vlsr);
	  else if (veldef == VEL_REL)
	    z2 = vrel2z(vlsr);
	  rf2 = x[i] * (1+z2);
	  dprintf(1,"0c: Line at %f %s has restfreq %f %s z=%g\n",x[i],xunit,rf2,xunit,z2);
	  // add freq
	  rfreq[nrfreq++] = rf2;
	}
      }
    }
  }
}

void line_id()
{
  int i, j, lmin=0;
  real dfmin, df, delta;
  char comment[3];

  if (ndvel > 0) {
    // need to check ndvel first, since skyfreq is hidden in rfreq for spectra
    printf("# restfreq=%g\n",restfreq);
    printf("# SkyFreq Vrad  Vrel   Vopt   z\n");
    printf("# GHz     km/s  km/s   km/s\n");
    for (i=0; i<ndvel; i++) {
      real v = dvel[i];
      printf("%g  %g %g %g  %g\n", mode==0? x[i] : rfreq[i], vopt2vrad(v), z2vrel(vopt2z(v)), v, vopt2z(v));
    }
    return;
  }
  
  if (nrfreq==0) {
    printf("# There are no lines to identify:\n");
    return;
  }
  
  printf("# There are %d restfreq's to identify:%s\n",
	 nrfreq, lnum==0 ? " (use linelist= to find their names)" : "");

  if (lnum > 0) {
    printf("# Commented lines do not fall within dv=%g km/s\n",dv);
    printf("#    guess    ID    LINE     dFreq    deltaV  NAME\n");
    printf("#    GHz            Ghz      GHz      km/s\n");
  }

  for (i=0; i<nrfreq; i++) {   // loop over all frequencies found
    if (lnum == 0) {
      printf("%f\n",rfreq[i]);
      continue;
    }
    if (dv < 0) {
      dfmin = -1.0;
      for (j=0; j<lnum; j++) {   // loop over linelist and find nearest match
	if (dfmin < 0) {
	  dfmin = ABS(lfreq[j] - rfreq[i]);
	  lmin = j;
	  continue;
	}
	df = ABS(lfreq[j] - rfreq[i]);
	if (df < dfmin) {
	  dfmin = df;
	  lmin = j;
	}
      }
      delta = (1-rfreq[i]/lfreq[lmin])*ckms;
      if (ABS(delta) > -dv)
	strcpy(comment,"#");
      else
	strcpy(comment," ");
      // @todo do we need comment?
      printf("%s %f   %d %f %f %g %s\n",comment,rfreq[i], lmin, lfreq[lmin], dfmin, delta, lname[lmin]);
    } else {
      // find all neighbors within given dv (in km/s)
      bool found = FALSE;
      for (j=0; j<lnum; j++) {   // loop over linelist and find nearest match
	delta = (1-rfreq[i]/lfreq[j])*ckms;
	if (ABS(delta) < dv) {
	  strcpy(comment," ");
	  printf("%s %f   %d %f %f %g %s\n",comment,rfreq[i], lmin, lfreq[j], dfmin, delta, lname[j]);
	  found = TRUE;
	}
      }
      if (!found) {
	printf("# %f   no lines found within %g km/s\n", rfreq[i], dv);
      }
      printf("#\n");
    }
  }
}


/* from: tablsqfit */
void do_peak(real *xpeak, real *xerr, real *ypeak)
{
    real mat[(MAXCOL+1)*(MAXCOL+1)], vec[MAXCOL+1], sol[MAXCOL+1], a[MAXCOL+2];
    int i, j, k, range=1;
    int order = 2;

    for (i=1, j=0; i<npt; i++)			/* find the peak, at j */
        if (y[j] < y[i]) j = i;
    if (j==0 || j==npt-1) {			/* handle edge cases */
        warning("Peak at the edge");
        printf("%g %g\n",x[j],y[j]);
    }
    if (range==2) {
    	if (j==1 || j==npt-2) {
    	    warning("Peak too close to edge");
    	}
    }

    lsq_zero(order+1, mat, vec);
    for (i=j-range; i<=j+range; i++) {
        a[0] = 1.0;
        for (k=0; k<order; k++) {
            a[k+1] = a[k] * (x[i]-x[j]);
        }
        a[order+1] = y[i];
        lsq_accum(order+1,mat,vec,a,1.0);
    }
    lsq_solve(order+1,mat,vec,sol);
    dprintf(1,"Poly2 fit near j=%d (%g,%g)\n",j+1,x[j],y[j]);
    dprintf(1,"Peak:x,y= %g %g\n",
            x[j] - sol[1]/(2*sol[2]),
	    sol[0]-sol[1]*sol[1]/(4*sol[2]));
    *xpeak = x[j] - sol[1]/(2*sol[2]);
    *xerr = 0.0;   // TBD
    *ypeak = sol[0]-sol[1]*sol[1]/(4*sol[2]);
}

// Possible routines from other codes if fancier peak detection is needed
// see tabpeak and ccdmom
//  peak_spectrum
//  peak_mom
//  peak_find
//  peak_assign
